### PR TITLE
fix(compiler-cli): incorrectly type checking calls to implicit template variables

### LINKED ISSUE #39634: ngFor over functions causes type error with Ivy
# 🐞 bug report

### Affected Package

@angular/common, maybe? It's a compile-time error using nothing special except `*ngFor`

### Is this a regression?

This appeared as I was upgrading from Angular 9.1.0, though @JoostK below (https://github.com/angular/angular/issues/39634#issuecomment-725341454) was able to reproduce it on the latest 9.0.x and 9.1.x.

### Description

When looping over an array of functions using *ngFor, a strange typing error occurs when calling that function in the template. For example in this template it gives an error like the following:

```
export class AppComponent  {
  functions = [() => 1, () => 2];
}

<p *ngFor="let fn of functions">
	{{ fn() }}
</p>

Property 'fn' does not exist on type '() => number'.
```

You'll see that it somehow is trying to find `fn` on the function itself? It works fine if you turn off Ivy.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-ivy-vttvch?file=src/app/app.component.html

## 🔥 Exception or Error
<pre><code>Property '__' does not exist on type '___'.</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 10.2.0
Node: 12.18.1
OS: win32 x64

Angular: 10.2.2
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1002.0
@angular-devkit/build-angular   0.1002.0
@angular-devkit/core            10.2.0
@angular-devkit/schematics      10.2.0
@angular/cdk                    10.2.7
@angular/cli                    10.2.0
@angular/material               10.2.7
@schematics/angular             10.2.0
@schematics/update              0.1002.0
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Anything else relevant?**
Must be using ivy

### PR DESCRIPTION
**Pull request by crisbeto**
Currently when we encounter an implicit method call (e.g. `{{ foo(1) }}`) and we manage to resolve its receiver to something within the template, we assume that the method is on the receiver itself so we generate a type checking code to reflect it. This assumption is true in most cases, but it breaks down if the call is on an implicit receiver and the receiver itself is being invoked. E.g.

```
<div *ngFor="let fn of functions">{{ fn(1) }}</div>
```

These changes resolve the issue by generating a regular function call if the method call's receiver is pointing to `$implicit`.

Fixes #39634.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 1]
```diff
@@ -1604,15 +1604,21 @@ class TcbExpressionTranslator {
...
-      const method = ts.createPropertyAccess(wrapForDiagnostics(receiver), ast.name);
-      addParseSpanInfo(method, ast.nameSpan);
+      const receiver = this.targetToExpression(target, ast.sourceSpan);
+
+      // If the receiver was resolved to `$implicit`, it means that the
+      // implicit variable itself is being invoked so we have to generate
+      // a regular function call, rather than invoking a method on the receiver.
+      const expression = (target.value === '$implicit' && target.name === ast.name) ?
+          wrapForDiagnostics(receiver) :
+          ts.createPropertyAccess(wrapForDiagnostics(receiver), ast.name);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:1618]: NIT: I would just put this block into the `targetToExpression()` helper function. Otherwise, I would rename `targetToExpression()` to `targetToReceiver()`.

**Review by petebacondarwin — APPROVED** [review]
Nice work @crisbeto - it is great when things can be fixed so simply. Thanks for picking up this issue and resolving it so quickly.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 1 above)
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:1618]: I think my fix in #33132 was incorrect, as the local template variable name is irrelevant and should not be reflected into the TCB. This is independent from `$implicit`, as e.g. the following should also work:

```ts
export interface Context {
  fn: () => void;
}
```

```html
<ng-template let-function="fn">
  {{ function() }}
</ng-template>
```

should become (in pseudocode)

```ts
var _t1: Context;
var _t2 = _t1.fn;
_t2();
```

The TCB would currently be `_t2.function()` but there is no `function` on the context.

tldr; I believe it's sufficient to only change `const method = wrapForDiagnostics(receiver)` , at which point refactoring the methids is no longer needed.

**Review by JoostK — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 1 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: You're totally right! I'll leave in my test, even though it's basically the same as the one you added in the earlier PR, because the syntax is a little different.

**Review by JoostK — APPROVED** [review]

**Review by gbumanzordev — APPROVED** [review]
